tidb-server/server/util.go

```diff
@@ -0,0 +1,342 @@
+// this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to
+// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"strconv"
+	"time"
+
+	mysql "github.com/pingcap/tidb/mysqldef"
+	"github.com/pingcap/tidb/util/arena"
+	"github.com/pingcap/tidb/util/hack"
+)
+
+func parseLengthEncodedInt(b []byte) (num uint64, isNull bool, n int) {
+	switch b[0] {
+	// 251: NULL
+	case 0xfb:
+		n = 1
+		isNull = true
+		return
+
+	// 252: value of following 2
+	case 0xfc:
+		num = uint64(b[1]) | uint64(b[2])<<8
+		n = 3
+		return
+
+	// 253: value of following 3
+	case 0xfd:
+		num = uint64(b[1]) | uint64(b[2])<<8 | uint64(b[3])<<16
+		n = 4
+		return
+
+	// 254: value of following 8
+	case 0xfe:
+		num = uint64(b[1]) | uint64(b[2])<<8 | uint64(b[3])<<16 |
+			uint64(b[4])<<24 | uint64(b[5])<<32 | uint64(b[6])<<40 |
+			uint64(b[7])<<48 | uint64(b[8])<<56
+		n = 9
+		return
+	}
+
+	// 0-250: value of first byte
+	num = uint64(b[0])
+	n = 1
+	return
+}
+
+func dumpLengthEncodedInt(n uint64) []byte {
+	switch {
+	case n <= 250:
+		return tinyIntCache[n]
+
+	case n <= 0xffff:
+		return []byte{0xfc, byte(n), byte(n >> 8)}
+
+	case n <= 0xffffff:
+		return []byte{0xfd, byte(n), byte(n >> 8), byte(n >> 16)}
+
+	case n <= 0xffffffffffffffff:
+		return []byte{0xfe, byte(n), byte(n >> 8), byte(n >> 16), byte(n >> 24),
+			byte(n >> 32), byte(n >> 40), byte(n >> 48), byte(n >> 56)}
+	}
+
+	return nil
+}
+
+func parseLengthEncodedBytes(b []byte) ([]byte, bool, int, error) {
+	// Get length
+	num, isNull, n := parseLengthEncodedInt(b)
+	if num < 1 {
+		return nil, isNull, n, nil
+	}
+
+	n += int(num)
+
+	// Check data length
+	if len(b) >= n {
+		return b[n-int(num) : n], false, n, nil
+	}
+
+	return nil, false, n, io.EOF
+}
+
+func dumpLengthEncodedString(b []byte, alloc arena.Allocator) []byte {
+	data := alloc.Alloc(len(b) + 9)
+	data = append(data, dumpLengthEncodedInt(uint64(len(b)))...)
+	data = append(data, b...)
+	return data
+}
+
+func dumpUint16(n uint16) []byte {
```

Comment:
maybe we can move these function to util/codec later,

e.g, AppendUint16(b []byte, n uint16) []byte